common/src/main/java/io/pravega/common/util/ByteArraySegment.java

```diff
@@ -291,6 +332,51 @@ public int readBytes(ByteArraySegment segment) {
             this.position += len;
             return len;
         }
+
+        @Override
+        public byte readByte() throws EOFException {
+            if (position >= ByteArraySegment.this.length) {
+                throw new EOFException();
+            }
+
+            byte result = ByteArraySegment.this.array[ByteArraySegment.this.startOffset + this.position];
+            this.position++;
+            return result;
+        }
+
+        @Override
+        public int readInt() throws EOFException {
+            int nextPos = this.position + Integer.BYTES;
+            if (nextPos > ByteArraySegment.this.length) {
+                throw new EOFException();
+            }
+            int r = BitConverter.readInt(ByteArraySegment.this.array, ByteArraySegment.this.startOffset + this.position);
```

Comment:
Here and below, if at construction time we create a ByteBuffer view of the array this can be faster.

Reply:
I have filed #4884 so that we may improve these core classes. This is getting a bit out of scope for this PR, which was to improve data allocation for Table APIs.